applications/FluidDynamicsApplication/custom_processes/shock_capturing_entropy_viscosity_process.cpp

```diff
@@ -0,0 +1,390 @@
+    {
+        UpdateNodalAreaProcess();
+    }
+
+    ComputeNodalEntropies();
+    ComputeArtificialMagnitudes();
+}
+
+
+void ShockCapturingEntropyViscosityProcess::UpdateNodalAreaProcess()
+{
+    CalculateNodalAreaProcess<false> nodal_area_process(mrModelPart);
+    nodal_area_process.Execute();
+}
+
+
+template<unsigned int WriteBufferIndex>
+void ShockCapturingEntropyViscosityProcess::ComputeNodalEntropies()
+{
+    if(mrModelPart.ElementsBegin() == mrModelPart.ElementsEnd()) return; // empty mpdelpart
+
+    const double heat_capacity_ratio = mrModelPart.ElementsBegin()->GetProperties().GetValue(HEAT_CAPACITY_RATIO);
+
+    block_for_each(mrModelPart.Nodes(), [heat_capacity_ratio](NodeType& r_node)
+    {
+        const double density = r_node.FastGetSolutionStepValue(DENSITY);
+        const double pressure = r_node.FastGetSolutionStepValue(PRESSURE);
+
+        const auto entropy = ComputeEntropy(density, pressure, heat_capacity_ratio);
+
+        r_node.FastGetSolutionStepValue(NUMERICAL_ENTROPY, WriteBufferIndex) = entropy;
+
+        r_node.SetValue(ARTIFICIAL_MASS_DIFFUSIVITY, 0.0);
+        r_node.SetValue(ARTIFICIAL_DYNAMIC_VISCOSITY, 0.0);
+        r_node.SetValue(ARTIFICIAL_CONDUCTIVITY, 0.0);
+        r_node.SetValue(ARTIFICIAL_BULK_VISCOSITY, 0.0);
+    });
+}
+
+
+void ShockCapturingEntropyViscosityProcess::ComputeArtificialMagnitudes()
+{
+    KRATOS_TRY
+
+    const double delta_time = mrModelPart.GetProcessInfo().GetValue(DELTA_TIME);
+
+    if(mrModelPart.ElementsBegin() == mrModelPart.ElementsEnd()) return; // Empty modelpart
+
+    const double heat_capacity_ratio = mrModelPart.ElementsBegin()->GetProperties().GetValue(HEAT_CAPACITY_RATIO);
+
+    block_for_each(mrModelPart.Elements(), [&](Element& r_element)
+    {
+        const auto inf_norm = ComputeElementalInfNormData(r_element, delta_time, heat_capacity_ratio);
+        const double h2 = ComputeHSquared(r_element);
+
+        r_element.SetValue(SHOCK_SENSOR, inf_norm.EntropyResidual);
+
+        const double mu_e = mEntropyConstant * h2 * inf_norm.Density * inf_norm.EntropyResidual;
+        const double mu_max = mEnergyConstant * std::sqrt(h2) * inf_norm.Density * inf_norm.TotalVelocity;
+        const double mu_h = std::min(mu_e, mu_max);
+
+        const double mu_rho = mArtificialMassDiffusivityPrandtl * mu_h / inf_norm.Density;
+        const double kappa  = mArtificialConductivityPrandtl * mu_h / (heat_capacity_ratio - 1.0);
+
+        DistributeVariablesToNodes(r_element, mu_h, mu_rho, kappa);
+    });
+
+    KRATOS_CATCH("")
+}
+
+
+void ShockCapturingEntropyViscosityProcess::DistributeVariablesToNodes(
+    Element& rElement,
+    const double ArtificialBulkViscosity,
+    const double ArtificialMassDiffusivity,
+    const double ArtificialConductivity) const
+{
+    auto& r_geometry = rElement.GetGeometry();
+    const double element_volume = r_geometry.LocalSpaceDimension() == 3 ? r_geometry.Volume() : r_geometry.Area();
```

Comment:
As a future improvement you can consider passing a lambda to do the check once (before the elemental loop).

Reply:
Done